artemis-protocols/artemis-amqp-protocol/src/main/java/org/apache/activemq/artemis/protocol/amqp/broker/AMQPMessage.java

```diff
@@ -1263,7 +1263,10 @@ public Long getLongProperty(String key) throws ActiveMQPropertyConversionExcepti
 
    @Override
    public Object getObjectProperty(String key) {
-      if (key.equals(MessageUtil.TYPE_HEADER_NAME.toString())) {
+
+      if (key.startsWith(AMQPMessageSupport.JMS_AMQP_MESSAGE_ANNOTATION_PREFIX)) {
```

Comment:
This doesn't seem right at all, as the specification reserves all Annotations Symbol and UnsignedLong keys for its own use but the broker seems to be populating annotations that violate that.  Any client or other intermediary would be within its rights (technically required) to close the connection upon receiving a message with these broker specific message annotations as per the spec: 

←
3.2.10 Annotations -  The annotations type is a map where the keys are restricted to be of type symbol or of type ulong. All ulong keys, and all symbolic keys except those beginning with "x-" are reserved. Keys beginning with "x-opt-" MUST be ignored if not understood. On receiving an annotation key which is not understood, and which does not begin with "x-opt", the receiving AMQP container MUST detach the link with a not-implemented error.

Reply:
@tabish121 what UnsignedLong keys? This is getObjectProperty().

This is about filtering. The issue is when moving a message to DLQ, the copy will apply a message annotation. 

Previously with AMQ5, users would be able to filter by JMS_AMQP_MA_ORIG_QUEUE. This is just bringing that possibility.

Reply:
@tabish121 getObjectProperty is only used on the server's side for filtering. This is just leverating what's already implemented through setAnnotation, and Message.referenceOriginalMessage

After the message is copied to DLQ or ExpiryQueue, this will allow what's in there to be filtered.

Reply:
You are putting illegal message annotations into the message from what I can tell, and the test seems to validate that.  The message that is received has annotations which are of the form ({_AMQ_ORIG_QUEUE=testExpiryThroughTTL, _AMQ_ORIG_ADDRESS=testExpiryThroughTTL, _AMQ_ORIG_MESSAGE_ID=69, _AMQ_ACTUAL_EXPIRY=1559769271365}) which is illegal according the spec section I posted and a receiver is technically required to close the connection if it sees them.

Reply:
@tabish121 these message annotations are already there and it is not part of this change.

These annotations are already taking place since 2017. If these are not valid, this should be the subject of further changes.. and nothing to do with this specific change.

Reply:
@tabish121 when moving to DLQ or Expiry, the message is copied..

we could (perhaps should) instead of using annotations here (I agree it's wrong), use a regular property.

After all when moving the message to DLQ or Expiry we are moving a copy of the message, and not the original message. (you could see it that way.. although that's one line of argument)


Or perhaps we could fix it properly, and use proper annotations. but that will break compatibility with previous AMQ5, as the message is always converted from OpenWire into AMQP. (AMQ5 never preserves the AMQP message on that sense)

Reply:
> we could (perhaps should) instead of using annotations here (I agree it's wrong), use a regular property.

You could use an annotation for this if you wanted, although one that wasn't illegal according to the spec would be my suggestion (I know crazy spec purists).  

> After all when moving the message to DLQ or Expiry we are moving a copy of the message, and not the original message. (you could see it that way.. although that's one line of argument)

Whether it is copied or not doesn't really come into play here as the annotation being used is illegal.  By supporting this selector feature you are just adding a finer point to the thing that the broker is doing that it really shouldn't be.  The broker could for instance store the value in an annotation named 'x-opt-original-destination' or some such and that would be legal.  It would however not be accessible to a client that doesn't have direct access to annotations but only to application properties.

> Or perhaps we could fix it properly, and use proper annotations. but that will break compatibility with previous AMQ5, as the message is always converted from OpenWire into AMQP. (AMQ5 never preserves the AMQP message on that sense)

I'm unclear how ActiveMQ 5 or openwire come into this as 5.x doesn't add the original destination set on expired and DLQ'd messages when it converts them to AMQP to the best of my knowledge.

Reply:
I was actually asked to do this because AMQ5 is supporting this, according to my user.

Reply:
@tabish121 what I meant in regard to the copy was to use regular application properties on the copy, rather than an annotation.

Reply:
I can only find one sneaky way to apply a selector that would allow you to match on the original destination when a message is in the DLQ when using an AMQP consumer with 5.x.  If the user is talking about an OpenWire consumer then that is a different thing entirely and you cannot just say the two are equivalent.  

Doing something more akin to what 5.x does is probably actually an easier to explain solution than the way you've done here as it doesn't involve knowledge of message annotations or even protocol for the most part (if tweaked a bit).  

In 5.x as it stands now you can create a consumer using Qpid JMS that is subscribed to the DLQ and provide a selector along the lines of "JMSDestination='queue://myOriginalQueue'" and that would match if the message had expired off the queue "myOriginalQueue".  This works because the property expression applied via the selector first checks if the message has an original destination stored in it and uses that instead of the current destination to match on the selector value.  If Artemis was to do something similarly clever then you wouldn't need to require the user to provide some arbitrary value like _AMQ_SOMETHING_OR_OTHER their code remains vendor and protocol neutral.

Reply:
Ok, i will use extraProperties and do something about that...

Just so you know this PR was about exposing the properties.. the use of messageAnnotations for this wasn't introduced here.


But i will fix it as part of this PR.


i will close it for now as I won't have time to finish it before 1 week. I may reopen it later.